### PR TITLE
ARROW-15718: [C++] Increase thread limit to work around thread issues

### PR DESCRIPTION
**Pull request by wjones127**
In #12339 we added one, which enabled joining one table to one dataset using `use_threads=false`. However, I found that joining two datasets hit the thread limit.

There are plans to find [a long-term fix](https://issues.apache.org/jira/browse/ARROW-16072) that can run these operations synchronously with fewer threads, but that won't be ready for the next release.

As a temporary fix for 8.0.0, I propose just bumping up the `local_states_` capacity.

### CODE REVIEW

File: cpp/src/arrow/compute/exec/hash_join.cc [hunk 1]
```diff
@@ -103,7 +103,8 @@ class HashJoinBasicImpl : public HashJoinImpl {
     filter_ = std::move(filter);
     output_batch_callback_ = std::move(output_batch_callback);
     finished_callback_ = std::move(finished_callback);
-    local_states_.resize(num_threads + 1);  // +1 for calling thread + worker threads
+    // Adding ten, since each side of join might have an IO thread being called from.
+    local_states_.resize(num_threads + 10);
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/hash_join.cc:107]:
```suggestion
    local_states_.resize(GetCpuThreadPoolCapacity() + io::GetIOThreadPoolCapacity() + 1);
```
This is slightly less arbitrary than 10 and we'd have to be doing something truly weird to go over this value.

File: cpp/src/arrow/compute/exec/hash_join.cc [hunk 2]
```diff
@@ -103,7 +103,8 @@ class HashJoinBasicImpl : public HashJoinImpl {
     filter_ = std::move(filter);
     output_batch_callback_ = std::move(output_batch_callback);
     finished_callback_ = std::move(finished_callback);
-    local_states_.resize(num_threads + 1);  // +1 for calling thread + worker threads
+    // Adding ten, since each side of join might have an IO thread being called from.
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/hash_join.cc:106]: Can you add `TODO(ARROW-15732)` to this comment so we know when to clean it up?

**Review by westonpace — COMMENTED** [review]
Thanks for realizing we still needed to patch this up.  This looks good to me but I think we can be just a touch safer (the cost of allocating these extra states should be insignificant).

**Comment by wjones127** [comment]
It looks like that Java failure is simply flaky right now. See other recent failures:

 * https://github.com/apache/arrow/runs/6061027109?check_suite_focus=true
 * https://github.com/apache/arrow/runs/6062605054?check_suite_focus=true

**Review by westonpace — APPROVED** [review]
Thanks for taking care of this.

**Pull request closed without merging**